File: 2019/py/day_10.py
import itertools as it
from math import atan2, gcd
from pathlib import Path


def main():
    text = (Path(__file__).parent / "../input/input_10.txt").read_text()
    grid = [list(s) for s in text.strip().splitlines()]
    w, h = len(grid[0]), len(grid)
    vectors = {v for v in it.product(range(-h, h), range(-w, w)) if gcd(*v) == 1}
    vectors = sorted(vectors, key=lambda v: -atan2(v[1], v[0]))

    def scan(loc: tuple[int, int], vaporize: bool) -> int:
        total = 0
        for dy, dx in it.cycle(vectors) if vaporize else vectors:
            y, x = loc[0] + dy, loc[1] + dx
            while 0 <= y < len(grid) and 0 <= x < len(grid[0]):
                if grid[y][x] == '#':
                    total += 1
                    if vaporize:
                        grid[y][x] = '.'
                        if total == 200:
                            return 100 * x + y
                    break
                y, x = y + dy, x + dx
        return total

    station = (0, 0)
    best = 0
    for y, x in it.product(range(h), range(w)):
        if grid[y][x] == '#' and (n := scan((y, x), False)) > best:
            station = (y, x)
            best = n
    print("Part 1:", best)
    print("Part 2:", scan(station, True))


if __name__ == "__main__":
    main()

File: 2019/rs/src/bin/day_10.rs
use gcd::Gcd;
use std::fs;

fn scan(grid: &[&[u8]], loc: (isize, isize), vectors: &[(isize, isize)], vaporize: bool) -> usize {
    let mut total = 0;
    for &(dy, dx) in vectors {
        let (mut y, mut x) = (loc.0 + dy, loc.1 + dx);
        while 0 <= y && (y as usize) < grid.len() && 0 <= x && (x as usize) < grid[0].len() {
            if grid[y as usize][x as usize] == b'#' {
                total += 1;
                if vaporize && total == 200 {
                    return (100 * x + y) as usize;
                }
                break;
            }
            y += dy;
            x += dx;
        }
    }
    total
}

fn main() {
    let text = fs::read_to_string("../input/input_10.txt").unwrap();
    let grid: Vec<_> = text.lines().map(|s| s.as_bytes()).collect();
    let (h, w) = (grid.len() as isize, grid[0].len() as isize);
    let mut vectors = vec![];
    for y in -h..h {
        for x in -w..w {
            if x.unsigned_abs().gcd(y.unsigned_abs()) == 1 {
                vectors.push((y, x));
            }
        }
    }
    vectors.sort_by(|&(y0, x0), &(y1, x1)| {
        let a = -(x0 as f32).atan2(y0 as f32);
        let b = -(x1 as f32).atan2(y1 as f32);
        a.partial_cmp(&b).unwrap()
    });
    let mut station = (0, 0);
    let mut best = 0;
    for y in 0..h {
        for x in 0..w {
            if grid[y as usize][x as usize] == b'#' {
                let n = scan(&grid, (y, x), &vectors, false);
                if n > best {
                    best = n;
                    station = (y, x);
                }
            }
        }
    }
    println!("Part 1: {}", best);
    println!("Part 2: {}", scan(&grid, station, &vectors, true));
}
